rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,1625 @@
+Kubernetes provides: AWS, GCP and Azure. Just because our users are most likely
+to be on one of those.
+
+So, the goal for integration tests is to somehow test Vector with Kubernetes
+versions from MSKV to latest, all the container runtimes listed above and,
+additionally, on AWS, GCP and Azure.
+
+We can combine our requirements with offers from cloud providers. For instance,
+`runhcs` (and Windows containers in general) are supported at Azure. Although,
+whether we want to address Windows containers support is a different topic, we
+still should plan ahead.
+
+We'll need to come up with an optimal configuration.
+
+##### Where to keep and how to manage integration infrastructure config
+
+This is a very controversial question.
+
+Currently we have:
+
+- the Vector repo (with the Github Actions based CI flow)
+- the test harness (also integrated with CI, but this is it's own thing)
+
+We don't necessarily have to choose one of those places: we can add a new
+location if it's justified enough.
+
+Let's outline the requirements on the properties of the solution:
+
+- We want to have the ability to run the checks from the Vector repo CI, i.e.
+  per commit, per PR, per tag etc. This might not be immediately utilized, but
+  we just want to have that option.
+
+- We want to consolidate the management of the _cloud_ resources we allocate and
+  pay for Kubernetes test infrastructure in a single place. This is to avoid
+  spreading the responsibility, duplicating the logic, reusing allocated
+  resources for all our testing needs, and simplify accounting and make the
+  configuration management more flexible.
+  We can, for example, have a shared dependency for Vector CI flow, Test Harness
+  invocations, locally run tests - and whatever else we have - to rely on.
+
+- We want our test infrastructure easily available for the trusted developers
+  (Vector core team) to run experiments and tests against locally. This doesn't
+  mean we want to automate this and include running tests locally against our
+  whole k8s test infrastructure - but the ability to do it with little effort is
+  very important: even if we employ super-reliable CI automation, the turnaround
+  time of going through it is way higher than conducting an experiment locally.
+  Locally means using local code tree and binaries - the infrastructure itself
+  is still in the cloud.
+
+- Ideally, we want the test system to be available not just to the Vector core
+  team, but to the whole open-source community. Of course, we don't want to give
+  unrestricted access to _our_ cloud testing infrastructure - but the solution
+  we employ should allow third-parties to bring their own resources. Things that
+  are local in essence (like `minikube`) should just work. There shouldn't be a
+  situation where one can't run tests in `minikube` because cloud parts aren't
+  available. We already have a similar constraints at the Vector Test Harness.
```

Comment:
Agree. Local tests are fine and we've been using the `docker-compose.yml` file to orchestrate this. But for external resource usage, that needs to be restricted to people on the Vector team for a myriad of reasons (security, cost, and so on).